src/channel_internal.c

```diff
@@ -0,0 +1,395 @@
+
+#include "c_pal/gballoc_hl.h"
+#include "c_pal/gballoc_hl_redirect.h"
+#include "c_pal/thandle.h"
+#include "c_pal/srw_lock.h"
+#include "c_pal/threadpool.h"
+
+#include "c_util/doublylinkedlist.h"
+#include "c_util/async_op.h"
+#include "c_util/rc_ptr.h"
+
+#include "../inc/c_util/channel_internal.h"
+
+typedef struct CHANNEL_INTERNAL_TAG
+{
+    THANDLE(THREADPOOL) threadpool;
+    SRW_LOCK_HANDLE lock;
+    DLIST_ENTRY op_list;
+}
+CHANNEL_INTERNAL;
+THANDLE_TYPE_DEFINE(CHANNEL_INTERNAL);
+
+typedef struct CHANNEL_OP_TAG
+{
+    PUSH_CALLBACK push_callback;
+    PULL_CALLBACK pull_callback;
+    void* push_context;
+    void* pull_context;
+    THANDLE(RC_PTR) data;
+    DLIST_ENTRY anchor;
+
+    THANDLE(CHANNEL_INTERNAL) channel_internal;
+    THANDLE(ASYNC_OP) async_op; // self reference to keep op alive even after user disposes
+
+    CHANNEL_CALLBACK_RESULT result;
+}CHANNEL_OP;
+
+
+static void execute_callbacks(void* context);
+
+static void channel_internal_dispose(CHANNEL_INTERNAL* channel_internal)
+{
+    /*Codes_SRS_CHANNEL_INTERNAL_43_099: [ channel_internal_dispose shall call srw_lock_destroy. ]*/
+    srw_lock_destroy(channel_internal->lock);
+
+    /*Codes_SRS_CHANNEL_INTERNAL_43_091: [ channel_internal_dispose shall release the reference to THANDLE(THREADPOOL). ]*/
+    THANDLE_ASSIGN(THREADPOOL)(&channel_internal->threadpool, NULL);
+}
+
+
+void channel_internal_close(THANDLE(CHANNEL_INTERNAL) channel_internal)
+{
+    CHANNEL_INTERNAL* channel_internal_ptr = THANDLE_GET_T(CHANNEL_INTERNAL)(channel_internal);
+
+    /*Codes_SRS_CHANNEL_INTERNAL_43_094: [ channel_internal_close shall call srw_lock_acquire_exclusive. ]*/
+    srw_lock_acquire_exclusive(channel_internal_ptr->lock);
+
+    /*Codes_SRS_CHANNEL_INTERNAL_43_095: [ channel_internal_close shall iterate over the list of pending operations and do the following: ]*/
+    for (DLIST_ENTRY* entry = DList_RemoveHeadList(&channel_internal_ptr->op_list); entry != &channel_internal_ptr->op_list; entry = DList_RemoveHeadList(&channel_internal_ptr->op_list))
+    {
+        CHANNEL_OP* channel_op = CONTAINING_RECORD(entry, CHANNEL_OP, anchor);
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_096: [ set the result of the operation to CHANNEL_CALLBACK_RESULT_ABANDONED. ]*/
+        channel_op->result = CHANNEL_CALLBACK_RESULT_ABANDONED;
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_097: [ call threadpool_schedule_work with execute_callbacks as work_function. ]*/
+        if (threadpool_schedule_work(channel_internal_ptr->threadpool, execute_callbacks, channel_op) != 0)
+        {
+            LogError("Failure in threadpool_schedule_work(execute_callbacks, channel_op)");
+        }
+    }
+
+    /*Codes_SRS_CHANNEL_INTERNAL_43_100: [ channel_internal_close shall call srw_lock_release_exclusive. ]*/
+    srw_lock_release_exclusive(channel_internal_ptr->lock);
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, THANDLE(CHANNEL_INTERNAL), channel_internal_create_and_open, THANDLE(THREADPOOL), threadpool)
```

Comment:
>threadpool [](http://example.com/codeflow?start=112&length=10)

is threadpool `NULL` a valid argument here?

Reply:
argument validation is performed at the `channel` module level, no need to perform again at `channel_internal` level. `channel_internal` is only meant to be used by `channel`.